dbos/system_database.go

```diff
@@ -1103,42 +1103,46 @@ func (s *systemDatabase) notificationListenerLoop(ctx context.Context) {
 const _DBOS_NULL_TOPIC = "__null__topic__"
 
 // Send is a special type of step that sends a message to another workflow.
-// Three differences with a normal steps: durability and the function run in the same transaction, and we forbid nested step execution
+// Can be called both within a workflow (as a step) or outside a workflow (directly).
+// When called within a workflow: durability and the function run in the same transaction, and we forbid nested step execution
 func (s *systemDatabase) Send(ctx context.Context, input WorkflowSendInput) error {
 	functionName := "DBOS.send"
 
-	// Get workflow state from context
+	// Get workflow state from context (optional for Send as we can send from outside a workflow)
 	wfState, ok := ctx.Value(workflowStateKey).(*workflowState)
-	if !ok || wfState == nil {
-		return newStepExecutionError("", functionName, "workflow state not found in context: are you running this step within a workflow?")
-	}
+	var stepID int
+	var isInWorkflow bool
 
-	if wfState.isWithinStep {
-		return newStepExecutionError(wfState.workflowID, functionName, "cannot call Send within a step")
+	if ok && wfState != nil {
+		isInWorkflow = true
+		if wfState.isWithinStep {
+			return newStepExecutionError(wfState.workflowID, functionName, "cannot call Send within a step")
+		}
+		stepID = wfState.NextStepID()
 	}
 
-	stepID := wfState.NextStepID()
-
 	tx, err := s.pool.Begin(ctx)
 	if err != nil {
 		return fmt.Errorf("failed to begin transaction: %w", err)
 	}
 	defer tx.Rollback(ctx)
 
-	// Check if operation was already executed and do nothing if so
-	checkInput := checkOperationExecutionDBInput{
-		workflowID: wfState.workflowID,
-		stepID:     stepID,
-		stepName:   functionName,
-		tx:         tx,
-	}
-	recordedResult, err := s.CheckOperationExecution(ctx, checkInput)
-	if err != nil {
-		return err
-	}
-	if recordedResult != nil {
-		// when hitting this case, recordedResult will be &{<nil> <nil>}
-		return nil
+	// Check if operation was already executed and do nothing if so (only if in workflow)
```

Comment:
Are we going to support temp workflows (for exactly-once semantics) in the future? Looks like if send is called outside of a workflow, it doesn't check for anything.

Reply:
For example, it'd be nice to support "here is an idempotency key, make sure we only send a message to a workflow exactly once".

Reply:
It raises the question: do we ever foresee a world where Send() should *not* be idempotent?

Reply:
Hmm I think making it idempotent is nice. The reason is that `send` will simply append a message to the message queue. If send is not idempotent, then the receiver would need to carefully deduplicate.